File: leetcode018.py
class Solution:
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        ls = len(nums)
        nums.sort()
        for i in range(ls - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j = i + 1
            while j < ls-2:
                while j > i+1 and j < ls -2 and nums[j] == nums[j - 1]:
                    j = j + 1
                k = j + 1
                l = ls-1
                while k < l:
                    curr = nums[i] + nums[j] + nums[k] + nums[l]
                    if curr == target:
                        res.append([nums[i], nums[j], nums[k], nums[l]])
                        while k < l and nums[k] == nums[k + 1]:
                            k = k + 1
                        while k < l and nums[l-1] == nums[l]:
                            l = l - 1
                        k = k + 1
                        l = l - 1
                    elif curr > target:
                        l = l - 1
                    else:
                        k = k + 1
                j = j + 1

        return res

if __name__ == "__main__":
    s = Solution()
    print(s.fourSum([-3,-4,-5,0,-5,-2,5,2,-3],3))

File: leetcodec++/leetcode018.cpp
//
// Created by 黄敏辉 on 2019-06-28.
//

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        if (nums.size() < 4)
            return res;
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 3; i++){
            int a = nums[i];
            if (i > 0 && nums[i-1] == a)
                continue;
            for (int j = i+1; j < nums.size() - 2; j++){
                int b = nums[j];
                if (j > i+1 && nums[j-1] == b)
                    continue;
                int l = j + 1, r = nums.size() - 1;
                while (l < r){
                    int c = nums[l], d = nums[r];
                    int val = a+b+c+d;
                    if (val == target) {
                        res.push_back(vector<int>({a, b, c, d}));
                        do{l++;}while(nums[l]==nums[l-1]&&l<r);
                        do{r--;}while(nums[r]==nums[r+1]&&l<r);
                    }
                    else if(val > target)
                        r--;
                    else
                        l++;
                }
            }
        }
        return res;
    }
};